### PR TITLE
[refactor] Decouple visual_bert embeddings forward function

### PR DESCRIPTION
**Pull request by suzyahyah**
Summary
- decouple `forward` into `encode_text` and `encode_image` functions for
readability and potential reusability.

Tested with

`mmf_run
datasets=hateful_memes
run_type=train_val_test
config=projects/hateful_memes/configs/visual_bert/from_coco.yaml
model=visual_bert
checkpoint.resume_zoo=visual_bert.finetuned.hateful_memes.from_coco
`

### CODE REVIEW

File: mmf/modules/embeddings.py [hunk 1]
```diff
@@ -322,22 +322,7 @@ def initialize_visual_from_pretrained(self):
...
-        image_text_alignment=None,
-    ):
-        """
-        input_ids = [batch_size, sequence_length]
-        token_type_ids = [batch_size, sequence_length]
-        visual_embedding = [batch_size, image_feature_length, image_feature_dim]
-        image_text_alignment = [batch_size, image_feature_length, alignment_dim]
-        """
-
+    def encode_text(self, input_ids, token_type_ids=None):
```
Comment (apsdehal) [inline mmf/modules/embeddings.py:325]: Can you add `*args, **kwargs` at the end to ensure easy forward compatibility?

File: mmf/modules/embeddings.py [hunk 2]
```diff
@@ -351,84 +336,138 @@ def forward(
...
-        if visual_embeddings is not None:
-            visual_embeddings = self.projection(visual_embeddings)
-            token_type_embeddings_visual = self.token_type_embeddings_visual(
-                visual_embeddings_type
+    def encode_image(
+        self,
+        visual_embeddings,
+        visual_embeddings_type=None,
+        position_embeddings_visual=None,
+        image_text_alignment=None,
```
Comment (apsdehal) [inline mmf/modules/embeddings.py:346]: Same here.

File: mmf/modules/embeddings.py [hunk 3]
```diff
@@ -351,84 +336,138 @@ def forward(
...
-                    position_ids_visual
-                )
+    def forward(
+        self,
+        input_ids,
+        token_type_ids=None,
+        visual_embeddings=None,
+        visual_embeddings_type=None,
+        position_embeddings_visual=None,
+        image_text_alignment=None,
```
Comment (apsdehal) [inline mmf/modules/embeddings.py:442]: Same here.

File: mmf/modules/embeddings.py [hunk 4]
```diff
@@ -351,84 +336,138 @@ def forward(
...
+            + position_embeddings_visual
+            + token_type_embeddings_visual
+        )
+        return v_embeddings
+
+    def get_position_embeddings_visual(
+        self,
+        visual_embeddings,
+        image_text_alignment=None,
+        position_embeddings_visual=None,
```
Comment (apsdehal) [inline mmf/modules/embeddings.py:374]: Same

**Review by apsdehal — APPROVED** [review]
Thanks for refactoring this. Minor comments, should be good to land after that.

**Comment by apsdehal** [comment]
Also, run black on the file you changed. CI Tests are failing.

File: mmf/modules/embeddings.py [hunk 5]
```diff
@@ -351,84 +336,138 @@ def forward(
...
+        # text embeddings
+        text_embeddings = self.encode_text(input_ids, token_type_ids=token_type_ids)
+
+        # visual embeddings
+        if visual_embeddings is not None:
+            v_embeddings = self.encode_image(
+                visual_embeddings,
+                visual_embeddings_type=visual_embeddings_type,
+                position_embeddings_visual=position_embeddings_visual,
+                image_text_alignment=image_text_alignment,
```
Comment (suzyahyah) [inline mmf/modules/embeddings.py:460]: Would *args and **kwargs need to be included here as well?

Comment (suzyahyah) [reply mmf/modules/embeddings.py:460]: And also where `get_position_embeddings_visual` and `encode_text` are called?

File: mmf/modules/embeddings.py (hunk 1 above)
Comment (suzyahyah) [reply mmf/modules/embeddings.py:325]: Thanks for the pointer. I have a qn about [where else it should go ](https://github.com/facebookresearch/mmf/pull/536/files/537fc81fd9a89c1f3b51fc020d55a2136d353b1b#r483316471)

File: mmf/modules/embeddings.py (hunk 5 above)
Comment (apsdehal) [reply mmf/modules/embeddings.py:460]: no, they only need to be added to signature not to the actual call.

Comment (suzyahyah) [reply mmf/modules/embeddings.py:468]: Anything else we need to do for this ? :)

Comment (apsdehal) [reply mmf/modules/embeddings.py:468]: Nope, once @vedanuj approves it internally, it should land automatically.

**Pull request closed without merging**